packages/astro/src/vite-plugins/state.ts

```diff
@@ -0,0 +1,39 @@
+/**
+ * A plugin that let users write components that can interact with the tutorial
+ * state, indirectly acting on the editor and WebContainer.
+ *
+ * The virtual module can be imported as:
+ *
+ * ```ts
+ * import { tutorialStore } from 'tutorialkit:state';
+ *
+ * export const MyComponent = () => (
+ *   <button onClick={() => tutorialStore.reset()}>Reset</button>
+ * );
+ * ```
+ */
+import type { VitePlugin } from '../types.js';
+import { fileURLToPath } from 'node:url';
+import path from 'node:path';
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+const virtualModuleId = 'tutorialkit:state';
+const resolvedVirtualModuleId = `${virtualModuleId}`;
+
+export const tutorialkitState: VitePlugin = {
+  name: 'tutorialkit-state-virtual-mod-plugin',
+  resolveId(id) {
+    if (id === virtualModuleId) {
+      return resolvedVirtualModuleId;
+    }
+  },
+  async load(id) {
+    if (id === resolvedVirtualModuleId) {
+      const pathToInit = path.join(__dirname, 'default', 'components', 'webcontainer.ts');
+
+      return `
```

Comment:
So if nothing is returned, I assume it just goes to the next loader?

Reply:
Yup! See https://vitejs.dev/guide/api-plugin.html#virtual-modules-convention for examples from the docs.